File: erlang/math/collatz.erl
%% Print the start below 1000 with the longest Collatz sequence, tracked in a fold over the range.
%% Run: escript collatz.erl

-module(collatz).
-export([main/1]).

chain_length(1) ->
    1;
chain_length(N) when N rem 2 =:= 0 ->
    1 + chain_length(N div 2);
chain_length(N) ->
    1 + chain_length(N * 3 + 1).

main(_) ->
    {Longest, Best} = lists:foldl(
        fun(Start, {_BestStart, BestLength} = Acc) ->
            case chain_length(Start) of
                Length when Length > BestLength -> {Start, Length};
                _ -> Acc
            end
        end,
        {1, 1},
        lists:seq(1, 999)
    ),
    io:format("~w ~w~n", [Longest, Best]).

File: ocaml/math/collatz.ml
(* Print the start below 1000 with the longest Collatz sequence, tracked in a fold over the range. *)
(* Run: ocaml collatz.ml *)

let limit = 1000

let chain_length start =
  let rec walk value length =
    if value = 1 then length
    else
      let next = if value mod 2 = 0 then value / 2 else (value * 3) + 1 in
      walk next (length + 1)
  in
  walk start 1

let () =
  let longest, best =
    List.fold_left
      (fun (longest, best) start ->
        let length = chain_length start in
        if length > best then (start, length) else (longest, best))
      (1, 1)
      (List.init (limit - 1) (fun i -> i + 1))
  in
  Printf.printf "%d %d\n" longest best
